lib/analyse.ml

```diff
@@ -256,48 +255,59 @@ module Analysis = struct
-    let cmd = ("", [| find; "."; "-maxdepth"; "3"; "-name"; "*.opam" |]) in
-    Current.Process.check_output ~cwd:dir ~cancellable:true ~job cmd
-    >>!= fun output ->
-    let opam_files =
-      String.split_on_char '\n' output
-      |> List.sort String.compare
-      |> List.filter_map (function
-           | "" -> None
-           | path ->
-               let path =
-                 if Astring.String.is_prefix ~affix:"./" path then
-                   Astring.String.with_range ~first:2 path
-                 else path
-               in
-               let consider_opam_file path =
-                 match Fpath.v path |> Fpath.segs with
-                 | [ _file ] -> true
-                 | segs when List.exists is_test_dir segs ->
-                     Current.Job.log job "Ignoring test directory %S" path;
-                     false
-                 | _ -> true
-               in
-               let full_path = Filename.concat (Fpath.to_string dir) path in
-               if is_empty_file full_path then (
-                 Current.Job.log job "WARNING: ignoring empty opam file %S" path;
-                 None)
-               else if consider_opam_file path then Some path
-               else None)
+    let ty = type_of_dir root in
+    let fold_on_opam_files () =
+      let module S = Set.Make (String) in
+      let opam_files path =
+        let test_pathv = Option.get (Fpath.rem_prefix root path) in
+        let test_path = Fpath.to_string test_pathv in
+        let consider_opam_file =
+          match Fpath.segs test_pathv with
+          | [] | _ :: [] -> true
+          | segs ->
+              if List.exists is_test_dir segs then (
+                Current.Job.log job "Ignoring test directory %S" test_path;
+                false)
+              else true
+        in
+        if is_empty_file (Fpath.to_string path) then (
+          Current.Job.log job "WARNING: ignoring empty opam file %S" test_path;
+          None)
+        else if consider_opam_file then Some test_path
+        else None
+      in
+      let is_opam_ext path = Ok (Fpath.has_ext "opam" path) in
+      let traverse path =
+        match Fpath.rem_prefix root path with
+        (* maxdepth=3 *)
+        | Some suffix ->
+            let is_max_depth_3 = List.length (Fpath.segs suffix) <= 3 in
+            Ok is_max_depth_3
+        | None ->
+            if Fpath.equal root path then Ok true
+            else
+              Fmt.error_msg "%S is not a prefix of %S" (Fpath.to_string root)
+                (Fpath.to_string path)
+      in
+      let add_opam_files path acc =
+        match opam_files path with None -> acc | Some path -> S.add path acc
+      in
+      Bos.OS.Path.fold ~elements:(`Sat is_opam_ext) ~traverse:(`Sat traverse)
+        add_opam_files S.empty [ root ]
+      |> Result.map S.elements
     in
+    Lwt_preemptive.detach fold_on_opam_files () >>!= fun opam_files ->
```

Comment:
Why did you choose to use `Lwt_preemptive` here?

Reply:
It was from previous code and related to this comment: https://github.com/ocurrent/ocaml-ci/pull/316#issuecomment-790468851